Investigate browser test failure in Happy Path

There seems to be an occasional browser test failure in the Happy Path browser test:

```
 ✖ Happy path (screenshots:
[2019-06-25T09:06:55.657Z]  /tmp/screenshots/2019-06-25_09-00-44/test-7/HeadlessChrome_75.0.3770_Linux_0.0.0/errors/1.png)
[2019-06-25T09:06:55.657Z] 
[2019-06-25T09:06:55.657Z]    1) AssertionError: expected 'Continue Submission' to deeply equal
[2019-06-25T09:06:55.657Z]       'Submitted'
[2019-06-25T09:06:55.657Z] 
[2019-06-25T09:06:55.657Z]       Browser: HeadlessChrome 75.0.3770 / Linux 0.0.0
[2019-06-25T09:06:55.657Z]       Screenshot:
[2019-06-25T09:06:55.657Z] 
[2019-06-25T09:06:55.657Z]    /tmp/screenshots/2019-06-25_09-00-44/test-7/HeadlessChrome_75.0.3770_Linux_0.0.0/errors/1.png
[2019-06-25T09:06:55.657Z] 
[2019-06-25T09:06:55.657Z]          122 |  await t
[2019-06-25T09:06:55.657Z]          123 |    .expect(dashboard.titles.textContent)
[2019-06-25T09:06:55.657Z]          124 |    .eql(manuscript.title)
[2019-06-25T09:06:55.657Z]          125 |    .expect(dashboard.statuses.textContent)
[2019-06-25T09:06:55.657Z]          126 |    // TODO this might cause a race condition
[2019-06-25T09:06:55.657Z]        > 127 |    .eql('Submitted')
[2019-06-25T09:06:55.657Z]          128 |})
[2019-06-25T09:06:55.657Z]          129 |
[2019-06-25T09:06:55.657Z]          130 |test('Ability to progress through the wizard is tied to
[2019-06-25T09:06:55.657Z]       validation', async t => {
[2019-06-25T09:06:55.657Z]          131 |  const navigationHelper = new NavigationHelper(t)
[2019-06-25T09:06:55.657Z]          132 |
[2019-06-25T09:06:55.657Z] 
[2019-06-25T09:06:55.657Z]          at eql (/home/xpub/test/submission.browser.js:127:6)
[2019-06-25T09:06:55.657Z] 
[2019-06-25T09:06:55.657Z] 
[2019-06-25T09:06:55.657Z]    2) Unhandled promise rejection:
[2019-06-25T09:06:55.657Z] 
[2019-06-25T09:06:55.657Z]       ExternalAssertionLibraryError { code: 'E53', isTestCafeError: true,
[2019-06-25T09:06:55.657Z]       callsite: CallsiteRecord { filename:
[2019-06-25T09:06:55.657Z]       '/home/xpub/test/submission.browser.js', lineNum: 126,
[2019-06-25T09:06:55.657Z]       callsiteFrameIdx: 6, stackFrames: [ [Object], [Object], [Object],
[2019-06-25T09:06:55.657Z]       [Object], [Object], [Object], [Object], CallSite {}, [Object],
[2019-06-25T09:06:55.657Z]       [Object], [Object], [Object], CallSite {} ], isV8Frames: true },
[2019-06-25T09:06:55.657Z]       errMsg: 'AssertionError: expected \'Continue Submission\' to deeply
[2019-06-25T09:06:55.657Z]       equal \'Submitted\'' }
[2019-06-25T09:06:55.657Z] 
[2019-06-25T09:06:55.657Z]       Browser: HeadlessChrome 75.0.3770 / Linux 0.0.0

```

Screenshot

![](https://alfred.elifesciences.org/job/pull-requests-projects/job/elife-xpub/job/PR-2174/1/artifact/build/screenshots/2019-06-25_09-00-44/test-7/HeadlessChrome_75.0.3770_Linux_0.0.0/errors/1.png)

--- a/packages/component-submission/client/pages/SubmissionWizard.js
+++ b/packages/component-submission/client/pages/SubmissionWizard.js
@@ -67,6 +67,7 @@ export const SubmissionWizard = ({
   const [currentStep, setCurrentStep] = useState(getCurrentStepFromPath())
   const [isUploading, setIsUploading] = useState(false)
   const [submissionAttempted, setsubmissionAttempted] = useState(false)
+  let updatePromise = Promise.resolve()
 
   const isLastStep = () => currentStep === STEP_NAMES.length - 1
   const initialValues = parseInputToFormData(data.manuscript)
@@ -78,15 +79,35 @@ export const SubmissionWizard = ({
     wizardSchema,
   ]
 
-  const handleSave = formValues =>
-    updateSubmission({
+  // save the promise from the submission update mutation
+  // the reason we are not storing using useState is that its asynchronous
+  const handleSave = formValues => {
+    updatePromise = updateSubmission({
       variables: { data: parseFormToOutputData(formValues) },
     })
 
+    updatePromise
+      .then(() => {
+        updatePromise = Promise.resolve()
+      })
+      .catch(() => {
+        updatePromise = Promise.resolve()
+      })
+
+    return updatePromise
+  }
+
+  // only submit once we are finished updating
+  // this is for the rare case that the form submit while its performing
+  // an auto save
   const handleSubmit = formValues =>
-    submitManuscript({
-      variables: { data: parseFormToOutputData(formValues) },
-    }).then(() => history.push(`/thankyou/${match.params.id}`))
+    updatePromise
+      .then(() =>
+        submitManuscript({
+          variables: { data: parseFormToOutputData(formValues) },
+        }),
+      )
+      .then(() => history.push(`/thankyou/${match.params.id}`))
 
   return (
     <Formik

